### DISCUSSION TITLE
How do I customize the jsonforms form component?

### DISCUSSION THREAD

**Discussion by yilia-jcq** [eclipse-emfcloud/emfcloud#141] [Coffee Editor]
As shown in the image below, my component does not have a "Valid" field and is not in tabular format, so where can I modify the code? Thank you～
![2022-03-15 14-11-42屏幕截图](https://user-images.githubusercontent.com/101612117/158318211-553e1183-b46d-4c4d-a075-fa6bad851404.png)

**Comment by eneufeld** [comment]
Hi @yilia-jcq ,
the valid field is there for validation purposes.
I assume your jsonschema looks like this:
```
properties: {
  comments: {
    type: 'array',
    items: {
      type: 'object',
      properties: {
        name: {type: 'string'},
        value: {type: 'string'}
      }
    }
  }
}
```

JSONForms automatically selects a table for this kind of schemas.
What you can do is provide a uischema which would allow you to define the ui the way you want it.
How to provide it depends on where the jsonforms component is. So are you using the properties view or the tree view?

**Reply by yilia-jcq** [reply]
Dear @eneufeId ,
Thank you very much for receiving your reply, maybe my expression is not perfect, if I want to modify the "Add to Comments" button wording, or delete the button, where should I go?

**Comment by yilia-jcq** [comment]
uischema.json
`{
            'type': 'Control',
            'scope': '#/properties/comments'
        }`

schema.json
`'autosarpackage': {
      '$id': '#autosarpackage',
            'title': 'AutosarPackage',
            'type': 'object',
            'properties': {
                'comments': {
                    'type': 'array',
                    'items': {
                        'type': 'object',
                        'properties': {
                            'name': {
                                'type': 'string'
                            },
                            'value': {
                                'type': 'string'
                            }
                        }
                    }
                }
            }
        }`

There is no "Valid" field in the above case

**Reply by eneufeld** [reply]
Hi @yilia-jcq ,
the valid field comes from jsonforms and is used to indicate whether a validation error exist in this row or not.
You can't simply rename or remove the button. In order to do so you need to provide a custom renderer see: https://jsonforms.io/docs/tutorial/custom-renderers

**Reply by yilia-jcq** [reply]
Dear @eneufeld ,
In the Coffee Editor project, where is the entry point to modify the custom jsonforms component? For example, the link you gave is written in react, so what is our project?

**Reply by eneufeld** [reply]
@yilia-jcq ,
How to provide the custom renderer depends on where the jsonforms component is. So are you using the properties view or the tree view?

**Reply by yilia-jcq** [reply]
Dear @eneufeld ,
tree view

**Reply by eneufeld** [reply]
@lucas-koehler can you help here out on how to provide a custom renderer?

**Reply by lucas-koehler** [reply]
Hello @yilia-jcq ,
to add custom renderers to the tree editor's detail, you need to extend the detail widget and rebind `DetailFormWidget` to your new widget.
To be more precise, the following steps are necessary:

- Extend [DetailFormWidget](https://github.com/eclipse-emfcloud/theia-tree-editor/blob/master/theia-tree-editor/src/browser/detail-form-widget.tsx) and override method `getJsonFormsConfig`. In this method you return a jsonforms config object that contains you custom renderers. This is probably done best by extending the `jsonFormsConfig` default config object. The `renderers` property contains the renderers to use. Make sure that your custom version includes the default renderers and your custom ones!
- To use your detail widget instead of the default one, you need to rebind it in the widget factory. This is done [here](https://github.com/eclipse-emfcloud/coffee-editor/blob/master/web/coffee-editor-extension/src/browser/coffee-editor-frontend-module.ts#L44-L49) right after the `treeContainer` was created. You can rebind with `treeContainer.rebind(DetailFormWidget).to(MyDetailWidget);`
- Now your custom widget should be used
- There's also a ticket to simplify customizing the detail widget in the tree editor in case you want to contribute: https://github.com/eclipse-emfcloud/theia-tree-editor/issues/20
